### PR TITLE
Suggestion: do not describe all properties twice

### PR DESCRIPTION
**Pull request by make-it-git**
May be it is enough to describe them only in $fields. $fields are self explanatory

### CODE REVIEW

File: src/BaseModel.php [hunk 1]
```diff
@@ -109,22 +116,33 @@ public function __set($key, $value)
...
+                if (array_key_exists('default', $this->fields[$key])) {
+                    // if we have some specific default value
+                    // other than null
+                    return $this->fields[$key]['default'];
+                }
+                // Accessing $obj->key when no default data is set returns null
+                // so we return it as default value for any described but not loaded property
+                return null;
+            }
+            return $this->fieldsData[$key];
```
Comment (darrynten) [inline src/BaseModel.php]: We can simplify, and flatten this a bit and reduce complexity

```php
public function __get($key)
{
    if (array_key_exists($key, $this->fields)) {
        if (array_key_exists('default', $this->fields[$key])) {
            // we can leave early
            return $this->fields[$key]['default'];
        }

        if (array_key_exists($key, $this->fieldsData)) {
            return $this->fieldsData[$key];
        }

        return null;
    }

    // normal exception handler
```

I have not tested but it should work

**Review by darrynten — CHANGES_REQUESTED** [review]
Reducing complexity in __get() suggestion

File: src/BaseModel.php [hunk 2]
```diff
@@ -258,24 +278,26 @@ private function getRemoteKey($localKey)
      */
     private function prepareObjectRow($key, $config)
     {
+        $value = $this->__get($key);
```
Comment (darrynten) [inline src/BaseModel.php:262]: Good call this 🍺

File: tests/SageOne/Models/AccountModelTest.php [hunk 3]
```diff
@@ -343,7 +332,10 @@ public function testSave()
 
         $accountModel->name = 'New Name';
 
-        $accountModel->save();
+        $expectedResponse = json_decode(file_get_contents(__DIR__ . '/../../mocks/Account/POST_Account_Save_RESP.json'));
+
+        $response = $accountModel->save();
+        $this->assertEquals($expectedResponse->ID, $response->ID);
         // (var_dump($accountModel->toJson()));
```
Comment (darrynten) [inline tests/SageOne/Models/AccountModelTest.php]: We can't have commented out code getting committed ;)

File: tests/SageOne/Models/ExampleModelTest.php [hunk 4]
```diff
@@ -357,9 +353,6 @@ public function testExamples()
         $data = json_decode(file_get_contents(__DIR__ . '/../../mocks/Example/GET_Example_Get_xx.json'));
         $exampleModel->loadResult($data);
 
-        // Expected lengths after loading
-        $this->assertCount(8, (array)$exampleModel);
-
```
Comment (darrynten) [inline tests/SageOne/Models/ExampleModelTest.php:362]: Any particular reason for these assertions to not still be there?

File: src/BaseModel.php [hunk 5]
```diff
@@ -109,22 +116,31 @@ public function __set($key, $value)
...
      */
     public function __get($key)
     {
         if (array_key_exists($key, $this->fields)) {
-            return $this->$key;
+            if (array_key_exists($key, $this->fieldsData)) {
+                // there is some data loaded so we return it
+                return $this->fieldsData[$key];
+            }
+            if (array_key_exists('default', $this->fields[$key])) {
```
Comment (darrynten) [inline src/BaseModel.php]: Where does the `default` key come from?

File: tests/SageOne/Models/ExampleModelTest.php (hunk 4 above)
Comment (make-it-git) [reply tests/SageOne/Models/ExampleModelTest.php:362]: This test will be rewritten with new methods, there is already this check.
This one checks not only fields, but also other properties

File: src/BaseModel.php (hunk 5 above)
Comment (make-it-git) [reply src/BaseModel.php]: I guess from there docs.
If key is nullable, we are allowed not to specify it because it is null by default

Comment (make-it-git) [reply src/BaseModel.php]: May be this is not required.
May be throw exception in __get if data is not loaded?

File: src/BaseModel.php (hunk 2 above)
Comment (make-it-git) [reply src/BaseModel.php:262]: Is there better way?

File: src/BaseModel.php (hunk 5 above)
Comment (darrynten) [reply src/BaseModel.php]: So it's maybe that we have potential `default` and `required` properties to add to the fields map?

File: src/BaseModel.php (hunk 2 above)
Comment (darrynten) [reply src/BaseModel.php:262]: I think that's the right way

File: tests/SageOne/Models/BaseModelTest.php [hunk 6]
```diff
@@ -189,23 +182,32 @@ protected function verifyAttributes(string $class, array $attributes)
...
+            $this->assertEquals(
+                $nullable,
+                $value[$name]['nullable'],
+                "Model {$className} Key {$name} Expected nullable to be {$nullableText} got {$nullableOptionText}"
+            );
+
+            $readonly = $options['readonly'];
+            $readonlyText = $readonly ? 'true' : 'false';
+            $readonlyOptionText = $value[$name]['readonly'] ? 'true' : 'false';
+
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php:205]: I think we add additional checks on min/max/regex etc here

File: src/Models/AccountCategory.php [hunk 7]
```diff
@@ -99,32 +53,32 @@ class AccountCategory extends BaseModel
         'id' => [
             'type' => 'integer',
             'nullable' => false,
-            'persistable' => true,
+            'readonly' => false,
```
Comment (darrynten) [inline src/Models/AccountCategory.php:56]: This now has a `required` on it if we look at the docs and our new rules

https://accounting.sageone.co.za/api/1.1.2/Help/ResourceModel?modelName=AccountCategory

**Review by darrynten — CHANGES_REQUESTED** [review]
So far so good, very minor requests

**Review by darrynten — APPROVED** [review]

**Pull request merged**